satpy/multiscene.py

```diff
@@ -45,14 +45,45 @@
 log = logging.getLogger(__name__)
 
 
-def stack(datasets):
-    """Overlay series of datasets on top of each other."""
+def stack(datasets, weights=None):
+    """Combine a series of datasets together.
+
+    On default, datasets are stacked on top of each other, so the last one is
+    on top. But if a sequence of weights arrays are provided the datasets will
+    be combined according to those weights. The result will be a composite
+    dataset where the data in each pixel is coming from the dataset having the
+    highest weight.
+
+    """
+    if weights:
+        return _stack_weighted(datasets, weights)
+
     base = datasets[0].copy()
     for dataset in datasets[1:]:
-        base = base.where(dataset.isnull(), dataset)
+        try:
+            base = base.where(dataset == dataset._FillValue, dataset)
+        except AttributeError:
+            base = base.where(dataset.isnull(), dataset)
     return base
 
 
+def _stack_weighted(datasets, weights):
+    """Stack datasets using weights."""
+    # Go through weights and set to zero where corresponding datasets have a value equals _FillValue or nan
+    for i, dataset in enumerate(datasets):
+        try:
+            weights[i] = xr.where(dataset == dataset._FillValue, 0, weights[i])
+        except AttributeError:
+            weights[i] = xr.where(dataset.isnull(), 0, weights[i])
+
+    indices = da.argmax(da.dstack(weights), axis=-1)
+    dims = datasets[0].dims
+    attrs = datasets[0].attrs
```

Comment:
Yes I think we briefly discussed it. It is not obvious to me what you should do to make it intuitive and right. Which is why I would prefer to keep it as simple as possible. It feels like it will depend on the use case. It maybe that some scenes are not contributing at all to the final product, though they have been part of the stacking/blending process. Sounds like we should perhaps make a function that does this blending of attributes, and that way at least it will be more explicit in the code for future adaptations/changes? Then I would suggest we keep it simple what that function then actually does for now. I would go for a special treatment of the scene start and end times, and take from either first or last for the rest of the attributes? What else, besides times need special care?

Reply:
https://github.com/pytroll/satpy/blob/9a496eb6d963b69e3ac50e11e4d7dd8f3ec37d3c/satpy/dataset/metadata.py#L30-L48

The function already exists ^. But it looks like it doesn't handle start/end time the way I thought it would. That special handling is apparently only in the internal reader code.

Reply:
Nice!

Now I have changed the tests so that the blended scene has the start and end times set to the oldest `start_time` and the newest `end_time` of the scenes in the weighted stack. And the test pass with this code in the stack function:

```
    attrs = combine_metadata(*[x.attrs for x in datasets])

    start_time = datetime.now()
    for s_time in [x.attrs['start_time'] for x in datasets if 'start_time' in x.attrs]:
        if s_time < start_time:
            start_time = s_time

    end_time = datetime.fromtimestamp(0)
    for e_time in [x.attrs['end_time'] for x in datasets if 'end_time' in x.attrs]:
        if e_time > end_time:
            end_time = e_time

    attrs['start_time'] = start_time
    attrs['end_time'] = end_time
```

I should of course put that time calculation in a separate function. But I suppose you would like to have a `combine_metadata` function that also handle the times as expected?
Here I am not averaging but looking for the time span that describes all the data resulting in the blended scene.

Also, do you think I should handle the case when a scene does not have a start and an end time, and how?
I would tend to require in that case that the user provides scenes that have start/end times! Or should all scenes have that (requirement of all readers)?

Reply:
@djhoese I feel like this is growing, so hope I can finish it soon and then we could add more complexity and more clever handling of attributes etc in other/future PRs when that becomes needed, does that sound reasonable?

Reply:
I am OK keeping this as the first Scene metadata for now. If you have the time I think `combine_metadata` would at least be more accurate. If you also have the time a future PR to combine_metadata to min/max start/end time would be appreciated. I know the conversation has come up in the past so hopefully I'm not forgetting some reason for why it isn't already that way.